src/functional/iterator/embedded.py

```diff
@@ -374,6 +374,10 @@ def unstructured_domain(*args: NamedRange) -> UnstructuredDomain:
 
 @builtins.named_range.register(EMBEDDED)
 def named_range(tag: Tag | Dimension, start: int, end: int) -> NamedRange:
+    if isinstance(start, ConstantField):
+        start = start.value
+    if isinstance(end, ConstantField):
+        end = end.value
```

Comment:
I forgot to mention yesterday but there is a simpler and cleaner way to avoid problems like this with `ConstantField`s: `ConstantField` should support conversion to scalars by implementing  `__int__` and `__float__`,  `__bool__` and  `__complex__` (throwing TypeErrors when its value does not match the required type). Then, in client code potentially accepting scalars or 0-d fields, it would be enough to do something like: 

```python
from typing import SupportsInt
...
assert isinstance(start, SupportsInt)
assert isinstance(end SupportsInt)
...
 range(int(start), int(end))
```

This is completely in line with the idea of 0d-fields being both fields and scalars, and even the typing can be nicely expressed by using `SupportsInt` (or `SupportsFloat`, ...) from `typing`.